system/Database/Query.php

```diff
@@ -275,35 +275,24 @@ protected function compileBinds()
     {
         $sql = $this->finalQueryString;
 
-        $hasBinds      = strpos($sql, $this->bindMarker) !== false;
-        $hasNamedBinds = ! $hasBinds
-            && preg_match('/:(?!=).+:/', $sql) === 1;
-
-        if (empty($this->binds)
-            || empty($this->bindMarker)
-            || (! $hasNamedBinds && ! $hasBinds)
-        ) {
+        if (empty($this->binds)) {
             return;
         }
 
-        if (! is_array($this->binds)) {
-            $binds     = [$this->binds];
-            $bindCount = 1;
-        } else {
-            $binds     = $this->binds;
+        $binds = is_array($this->binds) ? $this->binds : [$this->binds];
+
+        if (is_int(key(array_slice($binds, 0, 1)))) {
```

Comment:
Is `array_slice` necessary here? Since the array has just been created, the current index should be the first element, so I think just doing `is_int(key($binds))` is enough.

Reply:
Or using `array_key_first`?

Reply:
`$binds` may be a reference to `$this->binds`, and as PHP doesn't copy arrays until they are modified, the pointer position of `$binds` will be the same as of `$this->binds`, which may have been modifier beforehand, we can't know.

A benefit of `key(array_slice($binds, 0, 1))` over `reset($binds)` is that with the latter PHP would internally make a copy of the whole array, while the former just creates a tiny array of one item.

Good idea about [`array_key_first()`](https://www.php.net/manual/en/function.array-key-first.php). As we are targetting PHP 7.3+, that's probably what we want to use.